### PR TITLE
ILM: Add cluster update timeout on step retry

### PR DESCRIPTION
**Pull request by andreidan**
This adds a timeout when moving ILM back on to a failed step. In
case the master is struggling with processing the cluster update requests
these ones will expire (as we'll send them again anyway on the next ILM
loop run)

This also adds more descriptive source messages for the cluster state update
tasks to aid debugging.

### CODE REVIEW

**Comment by andreidan** [comment]
@elasticmachine update branch

**Comment by andreidan** [comment]
@elasticmachine update branch

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java [hunk 1]
```diff
@@ -204,33 +206,45 @@ void onErrorMaybeRetryFailedStep(String policy, IndexMetadata indexMetadata) {
...
+                    public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                        if (oldState.equals(newState) == false) {
+                            IndexMetadata newIndexMeta = newState.metadata().index(index);
+                            Step indexMetaCurrentStep = getCurrentStep(stepRegistry, policy, newIndexMeta);
+                            StepKey stepKey = indexMetaCurrentStep.getKey();
+                            if (stepKey != null && stepKey != TerminalPolicyStep.KEY && newIndexMeta != null) {
+                                logger.trace("policy [{}] for index [{}] was moved back on the failed step for as part of an automatic " +
+                                    "retry. Attempting to execute the failed step [{}] if it's an async action", policy, index, stepKey);
+                                maybeRunAsyncAction(newState, newIndexMeta, policy, stepKey);
+                            }
```
Comment (andreidan) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:244]: This is code formatting

**Comment by andreidan** [comment]
@elasticmachine update branch

**Review by DaveCTurner — COMMENTED** [review]
I like the more descriptive task sources.

Can we also/instead have a mechanism that more explicitly prevents two of these retries being enqueued for the same `(policy, index)` pair at the same time? This timeout sorta does so as long as the poll interval is greater than 30 seconds, but I think it'd be useful to give a hard guarantee of this.

Relatedly, I think the timeout should be (a function of) something like the ILM poll interval rather than hard-coded at 30 seconds. We do sometimes need to deal with clusters that simply cannot process cluster state updates in a reasonable time by extending timeouts.

**Comment by andreidan** [comment]
@elasticmachine update branch

**Comment by andreidan** [comment]
@DaveCTurner thanks for the suggestions and bringing this up. We've discussed this issue and we'll go ahead with an internal setting (non-documented) that controls this timeout as a first step and come back to it after we get a chance to review and figure out better heuristics on what the timeout, or maybe entire approach, should be. We want to make sure we don't add to the "cluster is overwhelmed" problem a "rollover couldn't be executed so there's now also a full disk" problem because the cluster state updates timed out. Using a setting will give users a chance to address this based on the situations as they observe them.

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java [hunk 2]
```diff
@@ -204,33 +205,45 @@ void onErrorMaybeRetryFailedStep(String policy, IndexMetadata indexMetadata) {
...
-                public void onFailure(String source, Exception e) {
-                    logger.error(new ParameterizedMessage("retry execution of step [{}] for index [{}] failed",
-                        failedStep.getKey().getName(), index), e);
-                }
+                    @Override
+                    public TimeValue timeout() {
+                        // we can afford to drop these requests if they timeout as on the next {@link
+                        // IndexLifecycleRunner#runPeriodicStep} run the policy will still be in the ERROR step, as we haven't been able
+                        // to move it back into the failed step, so we'll try again
+                        return LifecycleSettings.LIFECYCLE_STEP_MASTER_TIMEOUT_SETTING.get(clusterService.state().metadata().settings());
```
Comment (andreidan) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:218]: I think we can use the setting we already created to manipulate the ILM related master timeouts. What do you think @dakrone ?

Comment (dakrone) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:218]: That sounds reasonable to me :+1:

Comment (andreidan) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java:218]: @dakrone cool, thanks for confirming, this is ready for review then 🙏🏻

**Review by dakrone — APPROVED** [review]
LGTM, thanks for adding this Andrei!

**Pull request merged**